File: OsmAnd/src/net/osmand/plus/myplaces/tracks/TrackFiltersHelper.java
package net.osmand.plus.myplaces.tracks;

import androidx.annotation.NonNull;

import net.osmand.plus.OsmandApplication;
import net.osmand.plus.myplaces.tracks.filters.BaseTrackFilter;
import net.osmand.plus.myplaces.tracks.filters.DateTrackFilter;
import net.osmand.plus.myplaces.tracks.filters.FilterChangedListener;
import net.osmand.plus.myplaces.tracks.filters.ListTrackFilter;
import net.osmand.plus.myplaces.tracks.filters.OtherTrackFilter;
import net.osmand.plus.myplaces.tracks.filters.RangeTrackFilter;
import net.osmand.plus.myplaces.tracks.filters.TextTrackFilter;
import net.osmand.plus.myplaces.tracks.filters.TrackFilterType;

import java.util.Date;

public class TrackFiltersHelper {

	public static BaseTrackFilter createTextFilter(TrackFilterType filterType, FilterChangedListener listener) {
		return new TextTrackFilter(filterType, listener);
	}

	public static BaseTrackFilter createDateFilter(TrackFilterType trackFilterType, Long minDate, FilterChangedListener listener) {
		return new DateTrackFilter(trackFilterType, minDate, listener);
	}

	public static BaseTrackFilter createOtherFilter(OsmandApplication app, TrackFilterType trackFilterType, FilterChangedListener listener) {
		return new OtherTrackFilter(app, trackFilterType, listener);
	}

	public static BaseTrackFilter createSingleListFilter(OsmandApplication app, TrackFilterType trackFilterType, FilterChangedListener listener) {
		return new ListTrackFilter(app, trackFilterType, listener);
	}

	public static BaseTrackFilter createRangeFilter(OsmandApplication app, TrackFilterType trackFilterType, FilterChangedListener listener) {
		if (trackFilterType.getDefaultParams().size() < 2) {
			throw new IllegalArgumentException("RangeTrackFilter needs 2 default params: minValue, maxValue");
		}
		Object minValue = trackFilterType.getDefaultParams().get(0);
		Object maxValue = trackFilterType.getDefaultParams().get(1);
		if (minValue.getClass() != maxValue.getClass()) {
			throw new IllegalArgumentException("RangeTrackFilter's 2 default params (minValue, maxValue) must be the same type");
		}
		Class<?> parameterClass = trackFilterType.getProperty().getTypeClass();
		if (parameterClass == Double.class) {
			return new RangeTrackFilter<>((Double) minValue, (Double) maxValue, app, trackFilterType, listener);
		} else if (parameterClass == Float.class) {
			return new RangeTrackFilter<>((Float) minValue, (Float) maxValue, app, trackFilterType, listener);
		} else if (parameterClass == Integer.class) {
			return new RangeTrackFilter<>((Integer) minValue, (Integer) maxValue, app, trackFilterType, listener);
		} else if (parameterClass == Long.class) {
			return new RangeTrackFilter<>((Long) minValue, (Long) maxValue, app, trackFilterType, listener);
		}
		throw new IllegalArgumentException("Unsupported gpxParameter type class " + parameterClass);
	}

	public static BaseTrackFilter createFilter(OsmandApplication app, TrackFilterType trackFilterType, FilterChangedListener filterChangedListener) {
		BaseTrackFilter newFilter;

		switch (trackFilterType.getFilterType()) {
			case TEXT: {
				newFilter = createTextFilter(trackFilterType, filterChangedListener);
				break;
			}
			case RANGE: {
				newFilter = createRangeFilter(app, trackFilterType, filterChangedListener);
				break;
			}
			case DATE_RANGE: {
				newFilter = createDateFilter(trackFilterType, (new Date()).getTime(), filterChangedListener);
				break;
			}

			case OTHER: {
				newFilter = createOtherFilter(app, trackFilterType, filterChangedListener);
				break;
			}

			case SINGLE_FIELD_LIST: {
				newFilter = createSingleListFilter(app, trackFilterType, filterChangedListener);
				break;
			}

			default:
				throw new IllegalArgumentException("Unknown filterType " + trackFilterType);
		}
		return newFilter;
	}

	@NonNull
	public static Class<? extends BaseTrackFilter> getFilterClass(TrackFilterType trackFilterType) {
		Class<? extends BaseTrackFilter> filterClass = null;
		switch (trackFilterType.getFilterType()) {
			case TEXT: {
				filterClass = TextTrackFilter.class;
				break;
			}
			case RANGE: {
				filterClass = RangeTrackFilter.class;
				break;
			}
			case DATE_RANGE: {
				filterClass = DateTrackFilter.class;
				break;
			}

			case OTHER: {
				filterClass = OtherTrackFilter.class;
				break;
			}

			case SINGLE_FIELD_LIST: {
				filterClass = ListTrackFilter.class;
				break;
			}

			default:
				throw new IllegalArgumentException("Unknown filterType " + trackFilterType);
		}
		return filterClass;
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/gpx/filters/TrackFiltersHelper.kt
package net.osmand.shared.gpx.filters

import net.osmand.shared.extensions.currentTimeMillis
import net.osmand.shared.gpx.GpxParameter
import net.osmand.shared.gpx.filters.FilterType.DATE_RANGE
import net.osmand.shared.gpx.filters.FilterType.OTHER
import net.osmand.shared.gpx.filters.FilterType.RANGE
import net.osmand.shared.gpx.filters.FilterType.SINGLE_FIELD_LIST
import net.osmand.shared.gpx.filters.FilterType.TEXT
import kotlin.reflect.KClass

object TrackFiltersHelper {
	fun createTextFilter(
		filterType: TrackFilterType,
		listener: FilterChangedListener?): BaseTrackFilter {
		return TextTrackFilter(filterType, listener)
	}

	fun createDateFilter(
		trackFilterType: TrackFilterType,
		minDate: Long,
		listener: FilterChangedListener?): BaseTrackFilter {
		return DateTrackFilter(trackFilterType, minDate, listener)
	}

	fun createOtherFilter(
		trackFilterType: TrackFilterType,
		listener: FilterChangedListener?): BaseTrackFilter {
		return OtherTrackFilter(trackFilterType, listener)
	}

	fun createSingleListFilter(
		trackFilterType: TrackFilterType,
		listener: FilterChangedListener?): BaseTrackFilter {
		return if (trackFilterType === TrackFilterType.FOLDER) FolderTrackFilter(
			listener) else ListTrackFilter(trackFilterType, listener)
	}

	fun createRangeFilter(
		trackFilterType: TrackFilterType,
		listener: FilterChangedListener?): BaseTrackFilter {
		if (trackFilterType.defaultParams == null || trackFilterType.defaultParams.size < 2) {
			throw IllegalArgumentException("RangeTrackFilter needs 2 default params: minValue, maxValue")
		}
		val minValue: Any = trackFilterType.defaultParams[0]
		val maxValue: Any = trackFilterType.defaultParams[1]
		if (minValue::class != maxValue::class) {
			throw IllegalArgumentException("RangeTrackFilter's 2 default params (minValue, maxValue) must be the same type")
		}
		val parameter = trackFilterType.property
		if (isGpxParameterClass(parameter, Double::class)) {
			return RangeTrackFilter(
				minValue as Double,
				maxValue as Double,
				trackFilterType,
				listener)
		} else if (isGpxParameterClass(parameter, Float::class)) {
			return RangeTrackFilter(
				minValue as Float,
				maxValue as Float,
				trackFilterType,
				listener)
		} else if (isGpxParameterClass(parameter, Int::class)) {
			return RangeTrackFilter(
				minValue as Int,
				maxValue as Int,
				trackFilterType,
				listener)
		} else if (isGpxParameterClass(parameter, Long::class)) {
			return RangeTrackFilter(
				minValue as Long,
				maxValue as Long,
				trackFilterType,
				listener)
		}
		throw IllegalArgumentException("Unsupported gpxParameter type class " + parameter?.typeClass)
	}

	private fun isGpxParameterClass(parameter: GpxParameter?, javaClass: KClass<*>): Boolean {
		return parameter?.typeClass == javaClass
	}

	fun createFilter(
		trackFilterType: TrackFilterType,
		filterChangedListener: FilterChangedListener?): BaseTrackFilter {
		val newFilter: BaseTrackFilter = when (trackFilterType.filterType) {
			TEXT -> {
				createTextFilter(
					trackFilterType,
					filterChangedListener)
			}

			RANGE -> {
				createRangeFilter(
					trackFilterType,
					filterChangedListener)
			}

			DATE_RANGE -> {
				createDateFilter(
					trackFilterType,
					currentTimeMillis(),
					filterChangedListener)
			}

			OTHER -> {
				createOtherFilter(
					trackFilterType,
					filterChangedListener)
			}

			SINGLE_FIELD_LIST -> {
				createSingleListFilter(
					trackFilterType,
					filterChangedListener)
			}

			else -> throw IllegalArgumentException("Unknown filterType $trackFilterType")
		}
		return newFilter
	}

	fun getFilterClass(trackFilterType: TrackFilterType): KClass<out BaseTrackFilter> {
		val filterClass = when (trackFilterType.filterType) {
			TEXT -> {
				TextTrackFilter::class
			}

			RANGE -> {
				RangeTrackFilter::class
			}

			DATE_RANGE -> {
				DateTrackFilter::class
			}

			OTHER -> {
				OtherTrackFilter::class
			}

			SINGLE_FIELD_LIST -> {
				ListTrackFilter::class
			}

			else -> throw IllegalArgumentException("Unknown filterType $trackFilterType")
		}
		return filterClass
	}
}